### PR TITLE
Fix ##3479 ctrl+a select active cell output or preview markdown

### PR DESCRIPTION
**Pull request by YurongHe**
If no cell is active, Ctrl+A wouldn’t select anything global.

- Code cell is active
1) Mouse/cursor is in editor, Ctrl+A will select the code content.
2) Mouse/cursor is not in editor, Ctrl+A selects the output
image

- Markdown is active
1) In preview mode, Ctrl+A selects the text content.
2) In edit mode
    	Mouse/cursor is in editor, Ctrl+A selects the text in the editor
image

    	Mouse/cursor is not in editor, , Ctrl+A selects the text in the preview section
image

### CODE REVIEW

**Comment by chlafreniere** [comment]
After this change, can users still select output text without doing ctrl + a? That's my only concern.

File: src/sql/workbench/parts/notebook/cellViews/output.component.ts [hunk 1]
```diff
@@ -71,6 +95,14 @@ export class OutputComponent extends AngularDisposable implements OnInit {
 		}
 	}
 
+	@Input() set activeCellId(value: string) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/parts/notebook/cellViews/output.component.ts]: nit: Put this on a separate line so we're consistent style wise

File: src/sql/workbench/parts/notebook/cellViews/outputArea.component.ts [hunk 2]
```diff
@@ -41,6 +43,14 @@ export class OutputAreaComponent extends AngularDisposable implements OnInit {
 		}
 	}
 
+	@Input() set activeCellId(value: string) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/parts/notebook/cellViews/outputArea.component.ts]: nit: Put this on a separate line so we're consistent style wise

File: src/sql/workbench/parts/notebook/cellViews/textCell.css [hunk 3]
```diff
@@ -8,7 +8,11 @@ text-cell-component {
 }
 
 text-cell-component .notebook-preview {
-	user-select: initial;
+	user-select: none;
 	padding-left: 8px;
 	padding-right: 8px;
+}
+
+.notebook-preview.actionselect{
```
Comment (Charles-Gagnon) [inline src/sql/workbench/parts/notebook/cellViews/textCell.css]: nit: space

File: src/sql/workbench/parts/notebook/cellViews/outputArea.css [hunk 4]
```diff
@@ -10,4 +10,8 @@ output-area-component .notebook-output {
 	border-top-width: 0px;
 	user-select: text;
 	padding: 5px 20px 0px;
+}
+
+.output.actionselect {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/parts/notebook/cellViews/outputArea.css]: >.output.actionselect [](start = 0, length = 20)

You're adding the class "actionselect" in output.component.ts but here you're specifying the class as .output.actionselect - does that actually work as expected? Or is addClass doing something special?

Also please scope the classes as much as possible so the styles don't apply to things outside notebooks - something like this :

output-area-component .actionselect {

File: src/sql/workbench/parts/notebook/cellViews/output.component.html [hunk 5]
```diff
@@ -5,7 +5,7 @@
  *--------------------------------------------------------------------------------------------*/
 -->
 <div style="overflow: hidden; width: 100%; height: 100%; display: flex; flex-flow: column">
-	<div style="flex: 0 0 auto; user-select: initial;">
-		<div #output ></div>
+	<div style="flex: 0 0 auto; user-select: none;">
+		<div #output class="output" ></div>
```
Comment (Charles-Gagnon) [inline src/sql/workbench/parts/notebook/cellViews/output.component.html]: >output [](start = 22, length = 6)

output is pretty generic - I'd suggest giving it a more unique name. I would suggest notebook-output normally but that's taken by the output area...so maybe consider renmaing that one to notebook-output-area and this to notebook-output

File: src/sql/workbench/parts/notebook/cellViews/textCell.css [hunk 6]
```diff
@@ -8,7 +8,11 @@ text-cell-component {
...
+	user-select: text;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/parts/notebook/cellViews/textCell.css]: Also scope this to the component 

text-cell-component .notebook-preview.actionselect {

File: src/sql/workbench/parts/notebook/cellViews/output.component.ts [hunk 7]
```diff
@@ -14,8 +14,10 @@ import { MimeModel } from 'sql/workbench/parts/notebook/outputs/common/mimemodel
 import * as outputProcessor from 'sql/workbench/parts/notebook/outputs/common/outputProcessor';
 import { RenderMimeRegistry } from 'sql/workbench/parts/notebook/outputs/registry';
 import { IThemeService } from 'vs/platform/theme/common/themeService';
+import * as DOM from 'vs/base/browser/dom';
 
 export const OUTPUT_SELECTOR: string = 'output-component';
+export const USER_SELECT_CLASS ='actionselect';
```
Comment (Charles-Gagnon) [inline src/sql/workbench/parts/notebook/cellViews/output.component.ts]: >export [](start = 0, length = 6)

Don't really need to export this

File: src/sql/workbench/parts/notebook/cellViews/textCell.component.ts [hunk 8]
```diff
@@ -23,6 +24,7 @@ import { NotebookModel } from 'sql/workbench/parts/notebook/models/notebookModel
 import { CellToggleMoreActions } from 'sql/workbench/parts/notebook/cellToggleMoreActions';
 
 export const TEXT_SELECTOR: string = 'text-cell-component';
+export const USER_SELECT_CLASS ='actionselect';
```
Comment (Charles-Gagnon) [inline src/sql/workbench/parts/notebook/cellViews/textCell.component.ts]: >export [](start = 0, length = 6)

Don't really need to export this

File: src/sql/workbench/parts/notebook/cellViews/outputArea.css (hunk 4 above)
Comment (kevcunnane) [reply src/sql/workbench/parts/notebook/cellViews/outputArea.css]: 👍 Please do some additional scoping as suggested here.

File: src/sql/workbench/parts/notebook/cellViews/textCell.component.ts [hunk 9]
```diff
@@ -182,6 +185,17 @@ export class TextCellComponent extends CellView implements OnInit, OnChanges {
 		}
 	}
 
+	private toggleUserSelect(userSelect: boolean): void {
```
Comment (kevcunnane) [inline src/sql/workbench/parts/notebook/cellViews/textCell.component.ts]: Dumb question: why do we need to toggle this? Why can't it be on all the time?

**Comment by YurongHe** [comment]
> After this change, can users still select output text without doing ctrl + a? That's my only concern.

when do select, the cell will be active. so it works. I don't know how to select output without selecting cell. Is it possible?

File: src/sql/workbench/parts/notebook/cellViews/textCell.component.ts (hunk 9 above)
Comment (YurongHe) [reply src/sql/workbench/parts/notebook/cellViews/textCell.component.ts]: If it is on all the time, then all of them will be selected. This is the way I came up. Maybe there are better way to do it.

File: src/sql/workbench/parts/notebook/cellViews/output.component.html (hunk 5 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/parts/notebook/cellViews/output.component.html]: In fact why even add the class? Just use the ID for the CSS

---
In reply to: [274515156](https://github.com/Microsoft/azuredatastudio/pull/4981#discussion_r274515156) [](ancestors = 274515156)

**Pull request merged**